luxonis_train/nodes/heads/base_head.py

```diff
@@ -0,0 +1,77 @@
+from abc import abstractmethod
+from typing import Generic, Literal
+
+from luxonis_train.nodes.base_node import (
+    BaseNode,
+    ForwardInputT,
+    ForwardOutputT,
+)
+from luxonis_train.utils import deep_merge_dicts
+
+
+class BaseHead(
+    BaseNode[ForwardInputT, ForwardOutputT],
+    Generic[ForwardInputT, ForwardOutputT],
+):
+    """Base class for all heads in the model.
+
+    @type parser: str | None
+    @ivar parser: Parser to use for the head.
+    @type is_softmaxed: bool | None
+    @ivar is_softmaxed: Whether the head uses softmax or not.
+    """
+
+    parser: (
+        Literal[
+            "ClassificationParser",
+            "YOLO",
+            "YoloDetectionNetwork",
+            "SegmentationParser",
+        ]
+        | None
+    ) = None
+    is_softmaxed: bool | None = None
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+
+    def __init_subclass__(cls, **kwargs):
+        """Check if head has all required attributes."""
+        super().__init_subclass__()
+        if cls.parser is None:
+            raise ValueError(f"Head `{cls.__name__}` must define a parser.")
```

Comment:
Should we enforce specifying the parser here or it could be `None`? For example, we don't have a defined parser for `DiscSubNetHead`.

Reply:
I don't think we should enforce it. The NN archive support imho should be opt-in for new heads because it's only relevant if the user wants to use the model with dai. 

I think we should leave it `None` as default and also remove the `@abstractmethod` from `_get_custom_head_config`. Instead, it would raise `NotImplementedError` by default. In the `Core.archive` you'd check whether this method raises or not.
If it does, the archiving fails with a message like "Archiving failed, node {node_name} does not implement ...".
If it doesn't raise, only then we'd check that `parser` and `is_softmaxed` are not `None`.